runtime_scan/pkg/provider/gcp/client.go

```diff
@@ -0,0 +1,519 @@
+			// user did not specify zones, meaning he wants to scan all zones in that region
+			// first, get the region information from gcp
+			retRegion, err := c.regionsClient.Get(ctx, &computepb.GetRegionRequest{
+				Project: c.gcpConfig.ProjectID,
+				Region:  region.Name,
+			})
+			if err != nil {
+				_, err := handleGcpRequestError(err, "get region")
+				return nil, err
+			}
+
+			zones = append(zones, getZonesLastPart(retRegion.Zones)...)
+		} else {
+			// user specified specific zones to scan in that region
+			zones = append(zones, *region.Zones...)
+		}
+	}
+	return zones, nil
+}
+
+func (c *Client) listInstances(ctx context.Context, filter *string, zone string, scanScope models.GcpScanScope) ([]models.TargetType, error) {
+	var ret []models.TargetType
+
+	it := c.instancesClient.List(ctx, &computepb.ListInstancesRequest{
+		Filter:     filter,
+		MaxResults: utils.PointerTo[uint32](maxResults),
+		Project:    c.gcpConfig.ProjectID,
+		Zone:       zone,
+	})
+	for {
+		resp, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			_, err := handleGcpRequestError(err, "list vms")
+			return nil, err
+		}
+		if !isInScopeByTags(resp, scanScope.InstanceTagSelector, scanScope.InstanceTagExclusion) {
+			logrus.Debugf("Ignoring vm %v due to tags filters", *resp.Name)
+			continue
+		}
+		info, err := c.getVMInfoFromVirtualMachine(resp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get vminfo from virtual machine: %w", err)
+		}
+		ret = append(ret, info)
+	}
+
+	return ret, nil
+}
+
+func isInScopeByTags(vm *computepb.Instance, includeTags *[]models.Tag, excludeTags *[]models.Tag) bool {
+	if !hasIncludeTags(vm, includeTags) {
+		return false
+	}
+	if hasExcludeTags(vm, excludeTags) {
+		return false
+	}
+	return true
+}
+
+// AND logic - if tags = {tag1:val1, tag2:val2},
+// then a vm will be excluded/included only if it has ALL of these tags ({tag1:val1, tag2:val2}).
+func hasIncludeTags(vm *computepb.Instance, tags *[]models.Tag) bool {
+	if tags == nil {
+		return true
+	}
+	if len(*tags) == 0 {
+		return true
+	}
+	if len(vm.Tags.Items) == 0 {
+		return false
+	}
+
+	instanceTags := convertTagsToMap(vm.Tags)
+
+	for _, tag := range *tags {
+		val, ok := instanceTags[tag.Key]
+		if !ok {
+			return false
+		}
+		if !(strings.Compare(val, tag.Value) == 0) {
+			return false
+		}
+	}
+	return true
+}
+
+// AND logic - if tags = {tag1:val1, tag2:val2},
+// then a vm will be excluded/included only if it has ALL of these tags ({tag1:val1, tag2:val2}).
+func hasExcludeTags(vm *computepb.Instance, tags *[]models.Tag) bool {
+	if tags == nil {
+		return false
+	}
+	if len(*tags) == 0 {
+		return false
+	}
+	if len(vm.Tags.Items) == 0 {
+		return false
+	}
+
+	instanceTags := convertTagsToMap(vm.Tags)
+
+	for _, tag := range *tags {
+		val, ok := instanceTags[tag.Key]
+		if !ok {
+			return false
+		}
+		if !(strings.Compare(val, tag.Value) == 0) {
+			return false
+		}
+	}
+	return true
```

Comment:
No, here if tags are nil we return false (no exclude tags), this is because we don't want to exclude instance in the case that the user didn't provided any exclude tags
In hasIncludeTags, if tags are nil we return true, since no user tags mean it has include tags

Reply:
I meant L409-L420
The convert and the for loop with the logic inside

Reply:
I can move it into a function